File: code/07_regression_model.py
"""
阶段2：回归模型（住院天数预测）
功能：多元线性回归模型构建、评估与诊断
作者：成员B
"""

import pandas as pd
import numpy as np
import pickle
import warnings
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.stats.diagnostic import het_breuschpagan
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
import os

warnings.filterwarnings('ignore')
plt.rcParams['font.sans-serif'] = ['SimHei']  # 用于显示中文
plt.rcParams['axes.unicode_minus'] = False

# ==================== 1. 加载数据 ====================
print("=" * 60)
print("阶段2：回归模型开发（住院天数预测）")
print("=" * 60)

print("\n[1] 加载数据...")
with open('data/processed/data_splits.pkl', 'rb') as f:
    data_splits = pickle.load(f)

reg_data = data_splits['regression']
X_train = reg_data['X_train'].copy()
X_test = reg_data['X_test'].copy()
y_train = reg_data['y_train'].copy()
y_test = reg_data['y_test'].copy()
features_base = reg_data['features_base']
features_enhanced = reg_data['features_enhanced']

print(f"训练集: {len(X_train):,} 样本")
print(f"测试集: {len(X_test):,} 样本")

# ==================== 2. 特征编码 ====================
def encode_features(X, encoders=None, fit=True):
    """对分类变量进行编码"""
    X_encoded = X.copy()
    if encoders is None:
        encoders = {}
    
    # 识别分类变量
    categorical_cols = X.select_dtypes(include=['object']).columns.tolist()
    
    for col in categorical_cols:
        if fit:
            le = LabelEncoder()
            X_encoded[col] = le.fit_transform(X[col].astype(str))
            encoders[col] = le
        else:
            if col in encoders:
                # 处理未见过的类别
                X_encoded[col] = X[col].astype(str).apply(
                    lambda x: encoders[col].transform([x])[0] if x in encoders[col].classes_ else 0
                )
    
    return X_encoded, encoders

print("\n[2] 特征编码...")
X_train_encoded, encoders = encode_features(X_train, fit=True)
X_test_encoded, _ = encode_features(X_test, encoders=encoders, fit=False)

# ==================== 3. 基准模型 ====================
print("\n[3] 构建基准模型...")

# 使用基准特征集
X_train_base = X_train_encoded[features_base].copy()
X_test_base = X_test_encoded[features_base].copy()

# 标准化（可选，线性回归通常不需要，但有助于解释）
scaler_base = StandardScaler()
X_train_base_scaled = scaler_base.fit_transform(X_train_base)
X_test_base_scaled = scaler_base.transform(X_test_base)

# 训练模型
model_base = LinearRegression()
model_base.fit(X_train_base_scaled, y_train)

# 预测
y_train_pred_base = model_base.predict(X_train_base_scaled)
y_test_pred_base = model_base.predict(X_test_base_scaled)

# 评估
train_r2_base = r2_score(y_train, y_train_pred_base)
test_r2_base = r2_score(y_test, y_test_pred_base)
train_rmse_base = np.sqrt(mean_squared_error(y_train, y_train_pred_base))
test_rmse_base = np.sqrt(mean_squared_error(y_test, y_test_pred_base))
train_mae_base = mean_absolute_error(y_train, y_train_pred_base)
test_mae_base = mean_absolute_error(y_test, y_test_pred_base)

print("\n基准模型评估结果:")
print(f"训练集 - R²: {train_r2_base:.4f}, RMSE: {train_rmse_base:.4f}, MAE: {train_mae_base:.4f}")
print(f"测试集 - R²: {test_r2_base:.4f}, RMSE: {test_rmse_base:.4f}, MAE: {test_mae_base:.4f}")

# ==================== 4. 改进模型 ====================
print("\n[4] 构建改进模型...")

X_train_enhanced = X_train_encoded[features_enhanced].copy()
X_test_enhanced = X_test_encoded[features_enhanced].copy()

scaler_enhanced = StandardScaler()
X_train_enhanced_scaled = scaler_enhanced.fit_transform(X_train_enhanced)
X_test_enhanced_scaled = scaler_enhanced.transform(X_test_enhanced)

model_enhanced = LinearRegression()
model_enhanced.fit(X_train_enhanced_scaled, y_train)

y_train_pred_enhanced = model_enhanced.predict(X_train_enhanced_scaled)
y_test_pred_enhanced = model_enhanced.predict(X_test_enhanced_scaled)

train_r2_enhanced = r2_score(y_train, y_train_pred_enhanced)
test_r2_enhanced = r2_score(y_test, y_test_pred_enhanced)
train_rmse_enhanced = np.sqrt(mean_squared_error(y_train, y_train_pred_enhanced))
test_rmse_enhanced = np.sqrt(mean_squared_error(y_test, y_test_pred_enhanced))
train_mae_enhanced = mean_absolute_error(y_train, y_train_pred_enhanced)
test_mae_enhanced = mean_absolute_error(y_test, y_test_pred_enhanced)

print("\n改进模型评估结果:")
print(f"训练集 - R²: {train_r2_enhanced:.4f}, RMSE: {train_rmse_enhanced:.4f}, MAE: {train_mae_enhanced:.4f}")
print(f"测试集 - R²: {test_r2_enhanced:.4f}, RMSE: {test_rmse_enhanced:.4f}, MAE: {test_mae_enhanced:.4f}")

# ==================== 5. 模型诊断 ====================
print("\n[5] 模型诊断...")

# 使用改进模型进行诊断
residuals = y_test - y_test_pred_enhanced

# 5.1 残差正态性检验
shapiro_stat, shapiro_p = stats.shapiro(residuals[:5000])  # Shapiro-Wilk检验（样本量大时用前5000个）
print(f"\n残差正态性检验 (Shapiro-Wilk):")
print(f"  统计量: {shapiro_stat:.4f}, p值: {shapiro_p:.4f}")

# 5.2 多重共线性检查（VIF）
print(f"\n多重共线性检查 (VIF):")
vif_data = pd.DataFrame()
vif_data["Variable"] = features_enhanced
vif_data["VIF"] = [variance_inflation_factor(X_train_enhanced_scaled, i) 
                   for i in range(X_train_enhanced_scaled.shape[1])]
vif_data = vif_data.sort_values('VIF', ascending=False)
print(vif_data.head(10))
print(f"\nVIF > 10 的变量数: {(vif_data['VIF'] > 10).sum()}")

# ==================== 6. 结果可视化 ====================
print("\n[6] 生成可视化图表...")

os.makedirs('docs/figures', exist_ok=True)

# 6.1 模型系数条形图
fig, axes = plt.subplots(1, 2, figsize=(16, 6))

# 基准模型系数
coef_base = pd.DataFrame({
    'Feature': features_base,
    'Coefficient': model_base.coef_
})
coef_base = coef_base.sort_values('Coefficient', key=abs, ascending=False).head(15)

axes[0].barh(range(len(coef_base)), coef_base['Coefficient'])
axes[0].set_yticks(range(len(coef_base)))
axes[0].set_yticklabels(coef_base['Feature'])
axes[0].set_xlabel('回归系数')
axes[0].set_title('基准模型 - 主要变量系数')
axes[0].axvline(x=0, color='black', linestyle='--', linewidth=0.8)

# 改进模型系数
coef_enhanced = pd.DataFrame({
    'Feature': features_enhanced,
    'Coefficient': model_enhanced.coef_
})
coef_enhanced = coef_enhanced.sort_values('Coefficient', key=abs, ascending=False).head(15)

axes[1].barh(range(len(coef_enhanced)), coef_enhanced['Coefficient'])
axes[1].set_yticks(range(len(coef_enhanced)))
axes[1].set_yticklabels(coef_enhanced['Feature'])
axes[1].set_xlabel('回归系数')
axes[1].set_title('改进模型 - 主要变量系数')
axes[1].axvline(x=0, color='black', linestyle='--', linewidth=0.8)

plt.tight_layout()
plt.savefig('docs/figures/07_regression_coefficients.png', dpi=300, bbox_inches='tight')
print("✓ 已保存: docs/figures/07_regression_coefficients.png")

# 6.2 残差诊断图
fig, axes = plt.subplots(2, 2, figsize=(14, 10))

# 残差vs预测值
axes[0, 0].scatter(y_test_pred_enhanced, residuals, alpha=0.5, s=10)
axes[0, 0].axhline(y=0, color='r', linestyle='--')
axes[0, 0].set_xlabel('预测值')
axes[0, 0].set_ylabel('残差')
axes[0, 0].set_title('残差 vs 预测值')

# Q-Q图
stats.probplot(residuals, dist="norm", plot=axes[0, 1])
axes[0, 1].set_title('残差Q-Q图（正态性检验）')

# 残差直方图
axes[1, 0].hist(residuals, bins=50, edgecolor='black')
axes[1, 0].set_xlabel('残差')
axes[1, 0].set_ylabel('频数')
axes[1, 0].set_title('残差分布')

# 预测值vs实际值
axes[1, 1].scatter(y_test, y_test_pred_enhanced, alpha=0.5, s=10)
axes[1, 1].plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], 'r--', lw=2)
axes[1, 1].set_xlabel('实际值')
axes[1, 1].set_ylabel('预测值')
axes[1, 1].set_title(f'预测值 vs 实际值 (R²={test_r2_enhanced:.3f})')

plt.tight_layout()
plt.savefig('docs/figures/08_regression_diagnostics.png', dpi=300, bbox_inches='tight')
print("✓ 已保存: docs/figures/08_regression_diagnostics.png")

# ==================== 7. 保存模型 ====================
print("\n[7] 保存模型...")

os.makedirs('models', exist_ok=True)

model_info = {
    'model_base': model_base,
    'model_enhanced': model_enhanced,
    'scaler_base': scaler_base,
    'scaler_enhanced': scaler_enhanced,
    'encoders': encoders,
    'features_base': features_base,
    'features_enhanced': features_enhanced,
    'metrics_base': {
        'train_r2': train_r2_base,
        'test_r2': test_r2_base,
        'train_rmse': train_rmse_base,
        'test_rmse': test_rmse_base,
        'train_mae': train_mae_base,
        'test_mae': test_mae_base
    },
    'metrics_enhanced': {
        'train_r2': train_r2_enhanced,
        'test_r2': test_r2_enhanced,
        'train_rmse': train_rmse_enhanced,
        'test_rmse': test_rmse_enhanced,
        'train_mae': train_mae_enhanced,
        'test_mae': test_mae_enhanced
    },
    'vif_data': vif_data
}

with open('models/regression_model.pkl', 'wb') as f:
    pickle.dump(model_info, f)
print("✓ 已保存: models/regression_model.pkl")

# ==================== 8. 输出结果摘要 ====================
print("\n" + "=" * 60)
print("回归模型结果摘要")
print("=" * 60)
print(f"\n基准模型 (特征数: {len(features_base)}):")
print(f"  测试集 R²: {test_r2_base:.4f}")
print(f"  测试集 RMSE: {test_rmse_base:.4f}天")
print(f"  测试集 MAE: {test_mae_base:.4f}天")

print(f"\n改进模型 (特征数: {len(features_enhanced)}):")
print(f"  测试集 R²: {test_r2_enhanced:.4f}")
print(f"  测试集 RMSE: {test_rmse_enhanced:.4f}天")
print(f"  测试集 MAE: {test_mae_enhanced:.4f}天")

print(f"\n模型改进:")
print(f"  R²提升: {test_r2_enhanced - test_r2_base:.4f}")
print(f"  RMSE降低: {test_rmse_base - test_rmse_enhanced:.4f}天")

print("\n" + "=" * 60)
print("阶段2完成！回归模型已构建并评估。")
print("=" * 60)

File: code/07_regression_model.R
# ============================================================================
# 回归模型（住院天数预测）
# 功能：多元线性回归模型构建、评估与诊断
# ============================================================================

# 加载必要的包
library(dplyr)
library(caret)
library(car)          # VIF检查
library(ggplot2)
library(gridExtra)
library(stats)

# 设置中文字体（Windows）
if (.Platform$OS.type == "windows") {
  windowsFonts(SimHei = windowsFont("SimHei"))
  par(family = "SimHei")
}

# ==================== 1. 加载数据 ====================
cat(paste(rep("=", 60), collapse = ""), "\n")
cat("阶段2：回归模型开发（住院天数预测）\n")
cat(paste(rep("=", 60), collapse = ""), "\n")

cat("\n[1] 加载数据...\n")
data_splits <- readRDS('data/processed/data_splits.rds')

reg_data <- data_splits$regression
X_train <- reg_data$X_train
X_test <- reg_data$X_test
y_train <- reg_data$y_train
y_test <- reg_data$y_test
features_base <- reg_data$features_base
features_enhanced <- reg_data$features_enhanced

cat("训练集:", format(nrow(X_train), big.mark = ","), "样本\n")
cat("测试集:", format(nrow(X_test), big.mark = ","), "样本\n")

# ==================== 2. 特征编码 ====================
encode_features <- function(X, encoders = NULL, fit = TRUE) {
  # 对分类变量进行编码
  X_encoded <- X
  
  # 识别分类变量（字符型或因子型）
  categorical_cols <- names(X)[sapply(X, function(x) is.character(x) || is.factor(x))]
  
  if (is.null(encoders)) {
    encoders <- list()
  }
  
  for (col in categorical_cols) {
    if (fit) {
      # 训练时：创建编码器
      X_encoded[[col]] <- as.character(X[[col]])
      levels <- unique(X_encoded[[col]])
      X_encoded[[col]] <- as.numeric(factor(X_encoded[[col]], levels = levels))
      encoders[[col]] <- list(levels = levels)
    } else {
      # 测试时：使用已有编码器
      if (col %in% names(encoders)) {
        X_encoded[[col]] <- as.character(X[[col]])
        # 处理未见过的类别
        X_encoded[[col]] <- ifelse(
          X_encoded[[col]] %in% encoders[[col]]$levels,
          match(X_encoded[[col]], encoders[[col]]$levels) - 1,
          0
        )
      }
    }
  }
  
  return(list(X_encoded = X_encoded, encoders = encoders))
}

cat("\n[2] 特征编码...\n")
encoded_train <- encode_features(X_train, fit = TRUE)
X_train_encoded <- encoded_train$X_encoded
encoders <- encoded_train$encoders

encoded_test <- encode_features(X_test, encoders = encoders, fit = FALSE)
X_test_encoded <- encoded_test$X_encoded

# ==================== 3. 基准模型 ====================
cat("\n[3] 构建基准模型...\n")

# 使用基准特征集
X_train_base <- X_train_encoded[, features_base, drop = FALSE]
X_test_base <- X_test_encoded[, features_base, drop = FALSE]

# 标准化
preProc_base <- preProcess(X_train_base, method = c("center", "scale"))
X_train_base_scaled <- predict(preProc_base, X_train_base)
X_test_base_scaled <- predict(preProc_base, X_test_base)

# 准备数据框用于建模
train_data_base <- data.frame(X_train_base_scaled, y = y_train)
test_data_base <- data.frame(X_test_base_scaled, y = y_test)

# 构建公式
formula_base <- as.formula(paste("y ~", paste(features_base, collapse = " + ")))

# 训练模型
model_base <- lm(formula_base, data = train_data_base)

# 预测
y_train_pred_base <- predict(model_base, newdata = train_data_base)
y_test_pred_base <- predict(model_base, newdata = test_data_base)

# 评估指标
train_r2_base <- cor(y_train, y_train_pred_base)^2
test_r2_base <- cor(y_test, y_test_pred_base)^2
train_rmse_base <- sqrt(mean((y_train - y_train_pred_base)^2))
test_rmse_base <- sqrt(mean((y_test - y_test_pred_base)^2))
train_mae_base <- mean(abs(y_train - y_train_pred_base))
test_mae_base <- mean(abs(y_test - y_test_pred_base))

cat("\n基准模型评估结果:\n")
cat("训练集 - R²:", sprintf("%.4f", train_r2_base), 
    ", RMSE:", sprintf("%.4f", train_rmse_base), 
    ", MAE:", sprintf("%.4f", train_mae_base), "\n")
cat("测试集 - R²:", sprintf("%.4f", test_r2_base), 
    ", RMSE:", sprintf("%.4f", test_rmse_base), 
    ", MAE:", sprintf("%.4f", test_mae_base), "\n")

# ==================== 4. 改进模型 ====================
cat("\n[4] 构建改进模型...\n")

X_train_enhanced <- X_train_encoded[, features_enhanced, drop = FALSE]
X_test_enhanced <- X_test_encoded[, features_enhanced, drop = FALSE]

# 标准化
preProc_enhanced <- preProcess(X_train_enhanced, method = c("center", "scale"))
X_train_enhanced_scaled <- predict(preProc_enhanced, X_train_enhanced)
X_test_enhanced_scaled <- predict(preProc_enhanced, X_test_enhanced)

# 准备数据框
train_data_enhanced <- data.frame(X_train_enhanced_scaled, y = y_train)
test_data_enhanced <- data.frame(X_test_enhanced_scaled, y = y_test)

# 构建公式
formula_enhanced <- as.formula(paste("y ~", paste(features_enhanced, collapse = " + ")))

# 训练模型
model_enhanced <- lm(formula_enhanced, data = train_data_enhanced)

# 预测
y_train_pred_enhanced <- predict(model_enhanced, newdata = train_data_enhanced)
y_test_pred_enhanced <- predict(model_enhanced, newdata = test_data_enhanced)

# 评估指标
train_r2_enhanced <- cor(y_train, y_train_pred_enhanced)^2
test_r2_enhanced <- cor(y_test, y_test_pred_enhanced)^2
train_rmse_enhanced <- sqrt(mean((y_train - y_train_pred_enhanced)^2))
test_rmse_enhanced <- sqrt(mean((y_test - y_test_pred_enhanced)^2))
train_mae_enhanced <- mean(abs(y_train - y_train_pred_enhanced))
test_mae_enhanced <- mean(abs(y_test - y_test_pred_enhanced))

cat("\n改进模型评估结果:\n")
cat("训练集 - R²:", sprintf("%.4f", train_r2_enhanced), 
    ", RMSE:", sprintf("%.4f", train_rmse_enhanced), 
    ", MAE:", sprintf("%.4f", train_mae_enhanced), "\n")
cat("测试集 - R²:", sprintf("%.4f", test_r2_enhanced), 
    ", RMSE:", sprintf("%.4f", test_rmse_enhanced), 
    ", MAE:", sprintf("%.4f", test_mae_enhanced), "\n")

# ==================== 5. 模型诊断 ====================
cat("\n[5] 模型诊断...\n")

# 使用改进模型进行诊断
residuals <- y_test - y_test_pred_enhanced

# 5.1 残差正态性检验（Shapiro-Wilk，样本量大时用前5000个）
cat("\n残差正态性检验 (Shapiro-Wilk):\n")
n_test <- min(5000, length(residuals))
shapiro_test <- shapiro.test(residuals[1:n_test])
cat("  统计量:", sprintf("%.4f", shapiro_test$statistic), 
    ", p值:", sprintf("%.4f", shapiro_test$p.value), "\n")

# 5.2 多重共线性检查（VIF）
cat("\n多重共线性检查 (VIF):\n")
vif_values <- vif(model_enhanced)
vif_data <- data.frame(
  Variable = names(vif_values),
  VIF = as.numeric(vif_values)
)
vif_data <- vif_data[order(-vif_data$VIF), ]
print(head(vif_data, 10))
cat("\nVIF > 10 的变量数:", sum(vif_data$VIF > 10), "\n")

# ==================== 6. 结果可视化 ====================
cat("\n[6] 生成可视化图表...\n")

if (!dir.exists('docs/figures')) {
  dir.create('docs/figures', recursive = TRUE)
}

# 6.1 模型系数条形图
coef_base <- summary(model_base)$coefficients
coef_base_df <- data.frame(
  Feature = rownames(coef_base)[-1],  # 排除截距
  Coefficient = coef_base[-1, "Estimate"]
)
coef_base_df <- coef_base_df[order(abs(coef_base_df$Coefficient), decreasing = TRUE), ]
coef_base_df <- head(coef_base_df, 15)

coef_enhanced <- summary(model_enhanced)$coefficients
coef_enhanced_df <- data.frame(
  Feature = rownames(coef_enhanced)[-1],
  Coefficient = coef_enhanced[-1, "Estimate"]
)
coef_enhanced_df <- coef_enhanced_df[order(abs(coef_enhanced_df$Coefficient), decreasing = TRUE), ]
coef_enhanced_df <- head(coef_enhanced_df, 15)

p1 <- ggplot(coef_base_df, aes(x = reorder(Feature, Coefficient), y = Coefficient)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() +
  labs(x = "变量", y = "回归系数", title = "基准模型 - 主要变量系数") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(text = element_text(family = "SimHei", size = 10))

p2 <- ggplot(coef_enhanced_df, aes(x = reorder(Feature, Coefficient), y = Coefficient)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  coord_flip() +
  labs(x = "变量", y = "回归系数", title = "改进模型 - 主要变量系数") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "black") +
  theme_minimal() +
  theme(text = element_text(family = "SimHei", size = 10))

png('docs/figures/07_regression_coefficients.png', width = 16, height = 6, units = "in", res = 300)
grid.arrange(p1, p2, ncol = 2)
dev.off()
cat("✓ 已保存: docs/figures/07_regression_coefficients.png\n")

# 6.2 残差诊断图
png('docs/figures/08_regression_diagnostics.png', width = 14, height = 10, units = "in", res = 300)

par(mfrow = c(2, 2), family = "SimHei")

# 残差vs预测值
plot(y_test_pred_enhanced, residuals, 
     xlab = "预测值", ylab = "残差", 
     main = "残差 vs 预测值", pch = 20, cex = 0.5)
abline(h = 0, col = "red", lty = 2)

# Q-Q图
qqnorm(residuals, main = "残差Q-Q图（正态性检验）")
qqline(residuals, col = "red")

# 残差直方图
hist(residuals, breaks = 50, 
     xlab = "残差", ylab = "频数", 
     main = "残差分布", border = "black")

# 预测值vs实际值
plot(y_test, y_test_pred_enhanced, 
     xlab = "实际值", ylab = "预测值", 
     main = paste0("预测值 vs 实际值 (R²=", sprintf("%.3f", test_r2_enhanced), ")"),
     pch = 20, cex = 0.5)
abline(0, 1, col = "red", lty = 2, lwd = 2)

dev.off()
cat("✓ 已保存: docs/figures/08_regression_diagnostics.png\n")

# ==================== 7. 保存模型 ====================
cat("\n[7] 保存模型...\n")

if (!dir.exists('models')) {
  dir.create('models', recursive = TRUE)
}

model_info <- list(
  model_base = model_base,
  model_enhanced = model_enhanced,
  preProc_base = preProc_base,
  preProc_enhanced = preProc_enhanced,
  encoders = encoders,
  features_base = features_base,
  features_enhanced = features_enhanced,
  metrics_base = list(
    train_r2 = train_r2_base,
    test_r2 = test_r2_base,
    train_rmse = train_rmse_base,
    test_rmse = test_rmse_base,
    train_mae = train_mae_base,
    test_mae = test_mae_base
  ),
  metrics_enhanced = list(
    train_r2 = train_r2_enhanced,
    test_r2 = test_r2_enhanced,
    train_rmse = train_rmse_enhanced,
    test_rmse = test_rmse_enhanced,
    train_mae = train_mae_enhanced,
    test_mae = test_mae_enhanced
  ),
  vif_data = vif_data
)

saveRDS(model_info, 'models/regression_model.rds')
cat("✓ 已保存: models/regression_model.rds\n")

# ==================== 8. 输出结果摘要 ====================
cat("\n", paste(rep("=", 60), collapse = ""), "\n")
cat("回归模型结果摘要\n")
cat(paste(rep("=", 60), collapse = ""), "\n")
cat("\n基准模型 (特征数:", length(features_base), "):\n")
cat("  测试集 R²:", sprintf("%.4f", test_r2_base), "\n")
cat("  测试集 RMSE:", sprintf("%.4f", test_rmse_base), "天\n")
cat("  测试集 MAE:", sprintf("%.4f", test_mae_base), "天\n")

cat("\n改进模型 (特征数:", length(features_enhanced), "):\n")
cat("  测试集 R²:", sprintf("%.4f", test_r2_enhanced), "\n")
cat("  测试集 RMSE:", sprintf("%.4f", test_rmse_enhanced), "天\n")
cat("  测试集 MAE:", sprintf("%.4f", test_mae_enhanced), "天\n")

cat("\n模型改进:\n")
cat("  R²提升:", sprintf("%.4f", test_r2_enhanced - test_r2_base), "\n")
cat("  RMSE降低:", sprintf("%.4f", test_rmse_base - test_rmse_enhanced), "天\n")

cat("\n", paste(rep("=", 60), collapse = ""), "\n")
cat("阶段2完成！回归模型已构建并评估。\n")
cat(paste(rep("=", 60), collapse = ""), "\n")

